resharper/src/resharper-unity/CSharp/Daemon/Stages/Analysis/DrawGizmoAttributeProblemAnalyzer.cs

```diff
@@ -0,0 +1,83 @@
+using System;
+using JetBrains.Annotations;
+using JetBrains.ReSharper.Feature.Services.Daemon;
+using JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Errors;
+using JetBrains.ReSharper.Psi;
+using JetBrains.ReSharper.Psi.CSharp.Tree;
+using JetBrains.ReSharper.Psi.Util;
+using JetBrains.Util;
+using JetBrains.Util.Logging;
+
+namespace JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Analysis
+{
+    [ElementProblemAnalyzer(typeof(IAttribute),
+        HighlightingTypes = new[]
+        {
+            typeof(InvalidStaticModifierWarning), typeof(InvalidReturnTypeWarning),
+            typeof(InvalidTypeParametersWarning), typeof(InvalidParametersWarning)
+        })]
+    public class DrawGizmoAttributeProblemAnalyzer : MethodSignatureProblemAnalyzerBase<IAttribute>
+    {
+        private readonly IPredefinedTypeCache myPredefinedTypeCache;
+
+        public DrawGizmoAttributeProblemAnalyzer([NotNull] UnityApi unityApi, IPredefinedTypeCache predefinedTypeCache)
+            : base(unityApi)
+        {
+            myPredefinedTypeCache = predefinedTypeCache;
+        }
+
+        protected override void Analyze(IAttribute element, ElementProblemAnalyzerData data,
+            IHighlightingConsumer consumer)
+        {
+            if (!(element.TypeReference?.Resolve().DeclaredElement is ITypeElement attributeTypeElement))
+                return;
+
+            // Otherwise we'll treat it as targeting a method
+            if (element.Target == AttributeTarget.Return)
+                return;
+
+            if (attributeTypeElement.GetClrName().Equals(KnownTypes.DrawGizmo))
+            {
+                var methodDeclaration = MethodDeclarationNavigator.GetByAttribute(element);
+                CheckMethodDeclaration(methodDeclaration, element, consumer);
+            }
+        }
+
+        private void CheckMethodDeclaration(IMethodDeclaration methodDeclaration, IAttribute element, IHighlightingConsumer consumer)
+        {
+            if (methodDeclaration == null)
+                return;
+            
+            var predefinedType = myPredefinedTypeCache.GetOrCreatePredefinedType(element.GetPsiModule());
+            var gizmoType = TypeFactory.CreateTypeByCLRName(KnownTypes.GizmoType, predefinedType.Module);
+            var componentType = TypeFactory.CreateTypeByCLRName(KnownTypes.Component, predefinedType.Module);
+          
+
+            var expectedDeclaration = new MethodSignature(predefinedType.Void, true,
+                new[] {componentType, gizmoType},
+                new[] {"component", "gizmoType"});
+
+            var match = expectedDeclaration.Match(methodDeclaration);
+            
+            if (methodDeclaration.Params.ParameterDeclarations.Count == 2)
+            {
+                var parameters = methodDeclaration.Params.ParameterDeclarations;
+
+                if (parameters[0].Type.GetTypeElement()
+                        ?.IsDescendantOf(componentType.GetTypeElement()) == true)
```

Comment:
Looks like `IsDescendantOf` returns `true` if the derived type is equal to the base type. I don't think we want that - can you add a DrawGizmo for `MonoBehaviour` or `Component` directly?

Reply:
Unity doesn't appear to support gizmos on the base type of your class however if we add in that it is incorrect you'll kind of get stuck in an infinite cycle of it being wrong until you figure out that it can't be the base type. So we'll have to add an extra inspection for that I think.